### PR TITLE
instruments: Add Perfetto instrument

### PR DESCRIPTION
**Pull request by mrkajetanp**
This PR intends to add a Perfetto instrument that allows workload-automation to attach collecting a Perfetto trace to a running workload.

### CODE REVIEW

**Review by marcbonnici — CHANGES_REQUESTED** [review]

File: wa/instruments/perfetto.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        context.add_artifact('perfetto-bin', outfile, 'data')
+
+    def teardown(self, context):
+        path = self.target.path.join(self.target.working_directory, OUTPUT_PERFETTO_TRACE)
+        self.target.remove(path)
+
+    def finalize(self, context):
+        self.target.remove(self.target_config)
+
+    def validate(self, context):
```
[the remark this answers is not in the document]
Comment (mrkajetanp) [reply wa/instruments/perfetto.py]: all right, fixed

File: wa/instruments/perfetto.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    def finalize(self, context):
+        self.target.remove(self.target_config)
+
+    def validate(self):
+        if is_installed('trace-cmd'):
+            raise InstrumentError('perfetto cannot be used at the same time as trace-cmd')
+        if not os.path.isfile(self.config):
+            raise InstrumentError('perfetto config file not found at "{}"'.format(self.config))
+
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:102]: nit: trailing new line

File: wa/instruments/perfetto.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+
+from devlib import PerfettoCollector
+
+from wa import Instrument, Parameter
+from wa.framework import signal
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:21]: nit: Unused import

File: wa/instruments/perfetto.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        # push the config file to target
+        self.target.push(self.config, self.target_config)
+        collector_params = dict(
+            config=self.target_config,
+            force_tracebox=self.force_tracebox
+        )
+        self.collector = PerfettoCollector(self.target, **collector_params)
+
+    @very_slow
+    def start(self, context):
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:77]: `# pylint: disable=unused-argument`

File: wa/instruments/perfetto.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        self.collector.start()
+
+    @very_slow
+    def stop(self, context):
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:81]: `# pylint: disable=unused-argument`

File: wa/instruments/perfetto.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        self.collector.stop()
+
+    def update_output(self, context):
+        self.logger.info('Extracting Perfetto trace from target...')
+        outfile = os.path.join(context.output_directory, OUTPUT_PERFETTO_TRACE)
+        self.collector.set_output(outfile)
+        self.collector.get_data()
+        context.add_artifact('perfetto-bin', outfile, 'data')
+
+    def teardown(self, context):
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:91]: `# pylint: disable=unused-argument`

File: wa/instruments/perfetto.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        self.target.remove(self.collector.target_output_file)
+
+    def finalize(self, context):
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:94]: `# pylint: disable=unused-argument`

File: wa/instruments/perfetto.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                  Install tracebox even if traced is already running on the target device.
+                  If set to true, the tracebox binary needs to be placed in devlib's Package Bin directory.
+                  """)
+    ]
+
+    def __init__(self, target, **kwargs):
+        super(PerfettoInstrument, self).__init__(target, **kwargs)
+        self.collector = None
+
+    def initialize(self, context):
```
Comment (marcbonnici) [inline wa/instruments/perfetto.py:66]: `# pylint: disable=unused-argument`

**Review by marcbonnici — COMMENTED** [review]
LGTM, last few nits just to keep the linter [1] happy.
[1] https://github.com/ARM-software/workload-automation/actions/runs/6038173655/job/16561897216?pr=1203

**Comment by mrkajetanp** [comment]
Updated, the liner is now happy :)

**Review by marcbonnici — APPROVED** [review]

**Pull request merged**